pkg/apply/mutator/apply-time-mutator.go

```diff
@@ -0,0 +1,216 @@
+// Copyright 2021 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package mutator
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/cli-utils/pkg/apply/cache"
+	"sigs.k8s.io/cli-utils/pkg/kyq"
+	"sigs.k8s.io/cli-utils/pkg/object"
+	"sigs.k8s.io/cli-utils/pkg/object/mutation"
+)
+
+// ApplyTimeMutator mutates a resource by injecting values specified by the
+// apply-time-mutation annotation.
+// The optional ResourceCache will be used to speed up source reosurce lookups,
+// if specified.
+// Implements the Mutator interface
+type ApplyTimeMutator struct {
+	Client        dynamic.Interface
+	Mapper        meta.RESTMapper
+	ResourceCache cache.ResourceCache
+}
+
+// Name returns a mutator identifier for logging.
+func (atm *ApplyTimeMutator) Name() string {
+	return "ApplyTimeMutator"
+}
+
+// Mutate parses the apply-time-mutation annotation and loops through the
+// substitutions, applying each of them to the supplied target object.
+// Returns true with a reason, if mutation was performed.
+func (atm *ApplyTimeMutator) Mutate(obj *unstructured.Unstructured) (bool, string, error) {
```

Comment:
is there a test of this?

Reply:
not yet. it's on the TODO list, but I wanted to make sure I was on the righ track with impl and wasn't gonna have to refactor and write all new tests.

Reply:
added mutate tests:
https://github.com/karlkfi/cli-utils/blob/karl-apply-time-mutation/pkg/apply/mutator/apply_time_mutator_test.go

and kyq tests:
https://github.com/karlkfi/cli-utils/blob/karl-apply-time-mutation/pkg/kyq/kyq_test.go